### ISSUE TITLE
[BUG] Upgrade: longhorn-system can't be evicted

### ISSUE THREAD

**Issue by FrankYang0529** [harvester/harvester#2053]
**Describe the bug**
The upgrade progress is stuck and we can get error log in rancher pod.

```
evicting pod longhorn-system/instance-manager-r-262cb6bc
error when evicting pods/"instance-manager-r-262cb6bc" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
```

![Screen Shot 2022-03-21 at 5 00 28 PM](https://user-images.githubusercontent.com/4927361/159238119-1307d1a2-90ba-4583-8c79-fa521f5aeb33.png)

![Screen Shot 2022-03-21 at 4 59 38 PM](https://user-images.githubusercontent.com/4927361/159238230-0ba45fcc-814e-4223-a1a2-3c0e8932d5e5.png)

**To Reproduce**
1. Setup a 4-node v1.0.0 cluster.
2. Run `make default` to build latest image.
3. Create a Version CR for it.
```
apiVersion: harvesterhci.io/v1beta1
kind: Version
metadata:
  name: 1.0.2
  namespace: harvester-system
spec:
  isoChecksum: 4f80b9c5a728246c36c79f42ef8f1ad02e538b95b32d6f3dd70acdfeddb94808693371eff06dcc1af14d51876ed75ce72ff14926a883efe9251b11f9f3d5b10a
  isoURL: http://192.168.18.6:8000/harvester-7cc04b5-dirty-amd64.iso
  minUpgradableVersion: 0.3.0
  releaseDate: "20211231"
  tags:
  - dev
  - test
```
4. Create a Upgrade CR.
```
apiVersion: harvesterhci.io/v1beta1
kind: Upgrade
metadata:
  generateName: hvst-upgrade-
  namespace: harvester-system
spec:
  version: "1.0.2"
```
5. Upgrade process starts but can't finish.

**Expected behavior**
Upgrade process can finish.

**Support bundle**
[upgrade-log.zip](https://github.com/harvester/harvester/files/8314345/upgrade-log.zip)

**Environment:**
 - Harvester ISO version: `v1.0.0` to `master-370eca2a-head`
 - Underlying Infrastructure: KVM

https://github.com/harvester/harvester/issues/1861

**Comment by innobead** [comment]
If [allow-node-drain-with-the-last-healthy-replica](https://longhorn.io/docs/1.2.4/references/settings/#allow-node-drain-with-the-last-healthy-replica) is false by default, Longhorn will ensure at least one healthy volume replica available in another node before evicting all pods from the draining node. 

http://github.com/longhorn/longhorn-manager/blob/95b9b88e861f60ae7a39b396725530b953057e4c/controller/instance_manager_controller.go#L585-L619

cc @PhanLe1010

**Comment by TachunLin** [comment]
Encounter the similar issue during upgrade from v1.0.2 GA to v1.0.3-rc1. 

* One of the node stuck in the `Pre-drained` status
![image](https://user-images.githubusercontent.com/29251855/180755575-fe11b431-9a27-43cf-95d4-865ec3b700e9.png)

* Support bundle: 
[supportbundle_3a37fcaf-3ae7-4496-ae58-4efe0e3e7d61_2022-07-25T10-19-54Z.zip](https://github.com/harvester/harvester/files/9180257/supportbundle_3a37fcaf-3ae7-4496-ae58-4efe0e3e7d61_2022-07-25T10-19-54Z.zip)

### Test Information
* Test Environment: 4 nodes harvester on dolphin bare machine
* Harvester version: v1.0.2 GA to v1.0.3-rc1

**Comment by Vicente-Cheng** [comment]
The root cause is related to the issue https://github.com/longhorn/longhorn/issues/4305

**Comment by bk201** [comment]
We add the volumes health check in v1.1.0 before upgrading. Let's keep monitoring if this happens again.

**Comment by dseralathan** [comment]
node in pre-drained state 

 2022/12/06 21:38:50 [INFO] [planner] rkecluster fleet-local/local: waiting: draining bootstrap node(s) custom-3b6b250c5f97: draining node
2022/12/06 21:39:21 [INFO] [planner] rkecluster fleet-local/local: waiting: draining bootstrap node(s) custom-3b6b250c5f97: draining node
2022/12/06 21:40:11 [INFO] Downloading repo index from https://releases.rancher.com/server-charts/stable/index.yaml
2022/12/06 21:43:23 [INFO] Downloading repo index from http://harvester-cluster-repo.cattle-system/charts/index.yaml
2022/12/06 21:45:11 [INFO] Downloading repo index from https://releases.rancher.com/server-charts/stable/index.yaml
2022/12/06 21:47:50 [ERROR] Error during subscribe websocket: close sent
2022/12/06 21:48:23 [INFO] Downloading repo index from http://harvester-cluster-repo.cattle-system/charts/index.yaml
2022/12/06 21:48:51 [INFO] [planner] rkecluster fleet-local/local: waiting: draining bootstrap node(s) custom-3b6b250c5f97: draining node
2022/12/06 21:49:22 [INFO] [planner] rkecluster fleet-local/local: waiting: draining bootstrap node(s) custom-3b6b250c5f97: draining node

**Comment by shuo-wu** [comment]
The Longhorn version is v1.2.4.


The direct reason of the volume getting stuck and the only available replica (the gray one in the screenshot) cannot become mode `RW` is, the engine-controller in longhorn manager `longhorn-manager-vsv92` (on node `harvester-node-0`) cannot update the engine status:
```
2022-03-21T08:57:05.533058193Z E0321 08:57:05.532889       1 engine_controller.go:706] failed to update status for engine pvc-325c5123-eab0-4d14-899e-e9afdfcd8363-e-951e1c53: cannot get volume info: Timeout executing: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.2.4-rc1/longhorn [--url 10.52.0.112:10000 info], output , stderr, , error <nil>
```


In the corresponding engine manager log, it said that the engine was started:
```
2022-03-21T08:36:56.173496824Z [pvc-325c5123-eab0-4d14-899e-e9afdfcd8363-e-951e1c53] time="2022-03-21T08:36:56Z" level=info msg="Starting with replicas [\"tcp://10.52.1.22:10000\" \"tcp://10.52.2.23:10000\"]"

2022-03-21T08:36:56.202661406Z [pvc-325c5123-eab0-4d14-899e-e9afdfcd8363-e-951e1c53] time="2022-03-21T08:36:56Z" level=info msg="Adding backend: tcp://10.52.1.22:10000"
2022-03-21T08:36:56.205158455Z [pvc-325c5123-eab0-4d14-899e-e9afdfcd8363-e-951e1c53] time="2022-03-21T08:36:56Z" level=info msg="Adding backend: tcp://10.52.2.23:10000"

2022-03-21T08:36:58.200703776Z [pvc-325c5123-eab0-4d14-899e-e9afdfcd8363-e-951e1c53] time="2022-03-21T08:36:58Z" level=error msg="Backend tcp://10.52.2.23:10000 monitoring failed, mark as ERR: EOF"
2022-03-21T08:36:58.282925223Z [longhorn-instance-manager] time="2022-03-21T08:36:58Z" level=info msg="wait for gRPC service of process pvc-325c5123-eab0-4d14-899e-e9afdfcd8363-e-951e1c53 to start at localhost:10000"
2022-03-21T08:36:58.397689854Z [pvc-325c5123-eab0-4d14-899e-e9afdfcd8363-e-951e1c53] time="2022-03-21T08:36:58Z" level=info msg="Creating device /dev/longhorn/pvc-325c5123-eab0-4d14-899e-e9afdfcd8363 8:0"
...
```
Then there were lots of iscsi IO errors. I am not sure if these blocks the requests from the longhorn manager:
```
2022-03-21T08:37:12.392094102Z timeout_handler: Timeout request 0 due to disconnection
2022-03-21T08:37:12.408804161Z tgtd: bs_longhorn_request(105) fail to read at 0 for 4096
2022-03-21T08:37:12.408825428Z tgtd: bs_longhorn_request(149) io error 0x5566a0f39ab0 28 -14 4096 0, Success
......
```


BTW, the default engine image `longhornio/longhorn-engine:v1.2.4-rc1` are ready on all nodes.

**Comment by bk201** [comment]
@shuo-wu I think we mix two problems in the same issue. Vicente's comment https://github.com/harvester/harvester/issues/2053#issuecomment-1200829171 is for the support bundle in https://github.com/harvester/harvester/issues/2053#issuecomment-1193863957, if you have time please take a look again, thanks!

**Comment by shuo-wu** [comment]
OK. My last comment & analysis is for the original issue. I will check the comment you mentioned later.

**Comment by bk201** [comment]
Supposed to be fixed with LH 1.5.0: https://github.com/longhorn/longhorn/issues/4305

**Comment by markhillgit** [comment]
closing, expect that this is fixed in the new version of longhorn
